feat(benchmark infra) Add beforeEachBatchAsync callback

markfields opened the pull request:
## Description

Fixes [AB#19814](https://dev.azure.com/fluidframework/235294da-091d-4c29-84fc-cdfc3d90890b/_workitems/edit/19814)

This adds an async version of `beforeEachBatch`, to allow awaiting async code before running a batch of benchmark iterations.

## Reviewer Guidance

The merits of this design choice are up for debate, given that you can achieve this with the custom benchmark options.

markfields committed:
Add beforeEachBatchAsync

alexvy86 reviewed (commented):
IIRC there's a pattern somewhere in the benchmark package on how to use the sync VS async functions. I'm not sure I agree with just replacing `beforeEachBatch` with `beforeEachBatchAsync` withing `runBenchmarkAsync` is the right way to go. If the beforeEachBatch function is not async but the benchmarkFn is, I think it's still valid to use sync beforeEachBatch inside runBenchmarkAsync, unless we rewrite the contracts and the docs a bit, so `BenchmarkRunningOptionsAsync` can only take the async version of everything.

Depending on the final way we go, we might need to document this as a breaking change for the package (bump version to 0.52 and add an entry to CHANGELOG.md). Related: I think somewhere I might have a branch with the updated CHANGELOG that includes the changes in 0.51, I'll see if I can find it.

markfields commented:
> If the beforeEachBatch function is not async but the benchmarkFn is, I think it's still valid to use sync beforeEachBatch inside runBenchmarkAsync,

You're so right! I thought about this briefly then promptly forgot.  Should I run both beforeEachBatch's in the async case...?

markfields committed:
WIP: New approach / shuffling types a bit.

May have the same problem where existing calls
with sync beforeEachBatch + benchmarkFnAsync won't work.
Or maybe it does, but types don't block the converse.

markfields committed:
Ok I got something that actually works

markfields committed:
Fix tests

markfields committed:
api md

markfields committed:
Merge branch 'main' into benchmark-async-before

markfields committed:
Some fixes

markfields committed:
Prettier

markfields committed:
test cleanup

alexvy86 commented on tools/benchmark/src/test/Benchmark.unit.tests.ts:
```diff
@@ -54,6 +54,104 @@ describe("`benchmark` function", () => {
+				batches++;
+			},
+			benchmarkFnAsync: async (): Promise<void> => {
+				assert(batches > 0, "beforeEachBatchAsync should be called before test body");
```
```suggestion
				assert(batches > 0, "beforeEachBatch should be called before test body");
```

alexvy86 commented on tools/benchmark/src/test/Benchmark.unit.tests.ts:
```diff
@@ -54,6 +54,104 @@ describe("`benchmark` function", () => {
+				iterations++;
+			},
+			after: () => {
+				// Restore to actual batch count for 'after' hook logic
```
```suggestion
				// Restore to actual batch count for 'afterEach' hook logic which cares about the number of batches that ran
```

alexvy86 commented on tools/benchmark/src/test/Benchmark.unit.tests.ts:
```diff
@@ -54,6 +54,104 @@ describe("`benchmark` function", () => {
+				batches++;
+			},
+			benchmarkFn: (): void => {
+				assert(batches > 0, "beforeEachBatchAsync should be called before test body");
```
```suggestion
				assert(batches > 0, "beforeEachBatch should be called before test body");
```

alexvy86 commented on tools/benchmark/src/Configuration.ts:
```diff
@@ -209,20 +209,57 @@ export interface BenchmarkTimingOptions {
+ *
+ * @public
+ */
+export interface OnBatchAsync {
```
`OnBatch` and `OnBatchAsync` leave me a bit confused. Is it strictly necessary to have `beforeEachBatchAsync?: never;` in the former? And should `beforeEachBatch?: () => void;` really be deprecated on the latter? Seems like we allow both to be defined (since we call both in the "synthetic" `beforeAfterBatch` we craft inside `validateBenchmarkArguments`.

markfields replied:
The key feature is to block synchronous benchmark with asynchronous beforeEachBatch. I tried a number of strategies, this was the best I came up with.

If you can see a simpler way to keep the tests passing, I'm all for it. Or maybe that requirement I held to isn't worth it? Seems like it to me but open to discussion.

markfields replied:
I deprecated the one in the Async type because it's unnecessary, as you pointed out. Later someone could migrate usages to the new Async function and remove that one.

alexvy86 commented on tools/benchmark/src/Configuration.ts:
```diff
@@ -113,11 +112,9 @@ export type BenchmarkRunningOptions =
+export type BenchmarkRunningOptionsSync = BenchmarkSyncArguments;
 
-export type BenchmarkRunningOptionsAsync = BenchmarkAsyncArguments &
-	BenchmarkTimingOptions &
-	OnBatch;
+export type BenchmarkRunningOptionsAsync = BenchmarkAsyncArguments;
```
Nit: these two feel redundant now, maybe we should remove them?

markfields replied:
Fine with me. @CraigMacomber ?

CraigMacomber commented:
While I'm not generally opposed to this, I'd like to point out a couple things:

1. This feature should not be considered required to implement anything, since its perfectly possible to do everything it allows and more using `benchmarkFnCustom`. Note that even that is not our most flexible option: `benchmarkCustom` is even more flexible (can measure things other than runtime of batches)
2. A lot of benchmarks needing access to running code in-between batches that I have seen is very sketchy and was written with incorrect assumptions on how batches work (ex: if the benchmark function accumulates state that needs cleaning up, doing it between batches is a flawed approach since it leads to the amount of accumulated state being proportional to batch size which can impact performance (ex: GC speed) and the tests assume that the function being tested has performance independent of batch size. I have yet to see a good use of it that would not be more clearly and robustly done with `benchmarkFnCustom`. I don't particularly want to add more/better support and focus on this very commonly confusing and misused API.

CraigMacomber commented on tools/benchmark/api-report/benchmark.alpha.api.md:
```diff
@@ -172,6 +172,14 @@ export interface MochaExclusiveOptions {
 // @public
 export interface OnBatch {
     beforeEachBatch?: () => void;
+    beforeEachBatchAsync?: never;
+}
+
+// @public
+export interface OnBatchAsync {
+    // @deprecated
+    beforeEachBatch?: () => void;
+    beforeEachBatchAsync?: () => Promise<void>;
```
This change seems unnecessary complex. While we have to make the actual function being timed clearly sync or async to avoid the over head of checking which each iteration, I don't see an issue with making `beforeEachBatch?: () => void;` just typed as `beforeEachBatch?: () => void | promise<unknown>;` then just awaiting what ever it returns unconditionally.

That should keep the API identical, except that async beforeEachBatch will work. This is similar to how mocha deals with hooks: same API for sync and async.

markfields replied:
That's how I started.  I was bothered by the lack of compiler error when passing args with synchronous benchmarkFn by async beforeEachBatch

CraigMacomber replied:
Looking at the implementation, I guess that would be an issue with how a few things are currently factored, like `runBenchmarkSync`, so some of this refactoring might be needed to support async beforeEachBatch.

markfields commented:
Based on input from @CraigMacomber and @Abe27342 , closing this.  Quoting Abram:

> I tend to agree with Craig that most usages of beforeEachBatch aren't actually sound because they semantically want something more like beforeEachCycle, and using a custom benchmark mostly solves that use case (at the cost of not being as good for microbenchmarks and probably some others). I'd lean toward not bothering with it

My scenario is better covered via Custom Benchmark.
